chain-addr/src/lib.rs

```diff
@@ -300,40 +300,36 @@ fn is_valid_data(bytes: &[u8]) -> Result<(Discrimination, KindType), Error> {
 pub struct AddressReadable(String);
 
 impl AddressReadable {
-    const PRODUCTION_ADDRESS_PREFIX: &'static str = env!("PRODUCTION_ADDRESS_PREFIX");
-    const TEST_ADDRESS_PREFIX: &'static str = env!("TEST_ADDRESS_PREFIX");
-
     pub fn as_string(&self) -> &str {
         &self.0
     }
 
     /// Validate from a String to create a valid AddressReadable
-    pub fn from_string(s: &str) -> Result<Self, Error> {
+    pub fn from_string(expected_prefix: &str, s: &str) -> Result<Self, Error> {
         use std::str::FromStr;
         let r = Bech32::from_str(s)?;
-        let expected_discrimination = if r.hrp() == Self::PRODUCTION_ADDRESS_PREFIX {
-            Discrimination::Production
-        } else if r.hrp() == Self::TEST_ADDRESS_PREFIX {
-            Discrimination::Test
-        } else {
+        if r.hrp() != expected_prefix {
             return Err(Error::InvalidPrefix);
         };
         let dat = Vec::from_base32(r.data())?;
-        let (discrimination, _) = is_valid_data(&dat[..])?;
-        if discrimination != expected_discrimination {
-            return Err(Error::MismatchPrefix);
-        }
+        let _ = is_valid_data(&dat[..])?;
+
+        Ok(AddressReadable(s.to_string()))
+    }
+
+    pub fn from_string_anyprefix(s: &str) -> Result<Self, Error> {
```

Comment:
can we have a function to extract the _prefix_ too from the `AddressReadable`? This will make things useful for displaying a better error message to the users (and checking they are utilising the appropriate _prefix_ in a given context).